server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedBlobStore.java

```diff
@@ -57,28 +113,35 @@ public CachedBlobStore(BlobStoreCache cache, BlobStore backend,
     @Override
     public InputStream read(BucketName bucketName, BlobId blobId) throws ObjectStoreIOException, ObjectNotFoundException {
         return Mono.just(bucketName)
-            .filter(backend.getDefaultBucketName()::equals)
-            .flatMap(ignored -> Mono.from(cache.read(blobId))
-                    .<InputStream>flatMap(bytes -> Mono.fromCallable(() -> new ByteArrayInputStream(bytes))))
-            .switchIfEmpty(Mono.fromCallable(() -> backend.read(bucketName, blobId)))
+            .filter(getDefaultBucketName()::equals)
+            .flatMap(ignored -> readFromCache(blobId)
+                .flatMap(this::toInputStream))
+            .switchIfEmpty(readFromBackend(bucketName, blobId)
+                .flatMap(inputStream ->
+                    Mono.fromCallable(() -> ReadAheadInputStream.eager().of(inputStream).length(sizeThresholdInBytes))
+                        .flatMap(readAheadInputStream -> putInCacheIfNeeded(bucketName, readAheadInputStream, blobId)
+                            .then(Mono.fromCallable(() -> readAheadInputStream.in)))))
```

Comment:
why it's `then(fromCallable` and not `thenReturn` ?

Reply:
Yes, it work too.